perceval/components/processor.py

```diff
@@ -53,27 +53,57 @@ class Processor(AProcessor):
         >>> p = Processor("SLOS", BS() // PS() // BS())
 
     :param source: the Source used by the processor (defaults to perfect source)
+    :param noise: a NoiseModel containing noise parameters (defaults to no noise)
+                  Note: source and noise are mutually exclusive
     :param name: a textual name for the processor (defaults to "Local processor")
     """
-    def __init__(self, backend: Union[ABackend, str], m_circuit: Union[int, ACircuit] = None, source: Source = Source(),
-                 name: str = None):
+    def __init__(self, backend: Union[ABackend, str], m_circuit: Union[int, ACircuit] = None, source: Source = None,
+                 noise: NoiseModel = None, name: str = "Local processor"):
         super().__init__()
-        self._source = source
-        self.name = "Local processor" if name is None else name
+        self._init_backend(backend)
+        self._init_circuit(m_circuit)
+        self._init_noise(source, noise)
+        self.name = name
+        self._inputs_map: Union[SVDistribution, None] = None
+        self._simulator = None
 
+    def _init_noise(self, source: Source, noise: NoiseModel):
+        # Default = perfect simulation
+        self._phase_quantization = 0
+        self._source = Source()
+
+        # Backward compatibility case: the user passes a Source
+        if source is not None:
+            # If he also passed noise parameters: conflict between noise parameters => raise an exception
+            if noise is not None:
+                raise ValueError("Both 'source' and 'noise' parameters were set. You should only input a NoiseModel")
+            self._source = source
+
+        # The user passes a NoiseModel
+        elif noise is not None:
+            self.noise = noise
+
+    @AProcessor.noise.setter
+    def noise(self, nm):
+        super(Processor, type(self)).noise.fset(self, nm)
+        self._source = Source.from_noise_model(nm)
+        self._phase_quantization = nm.phase_imprecision
+        if isinstance(self._input_state, BasicState):
+            self._generate_noisy_input()
```

Comment:
This is cool to have thought about handle the source, but don't we need to do something about phase imprecision also ? Like rebuilding the circuit ?

Reply:
The method `linear_circuit` is handling the phase imprecision parameter (that's why I've overridden this method in `Processor`). I've added a docstring to this implementation so that it's clearer.